site/en/blog/what-is-the-top-layer/index.md

```diff
@@ -0,0 +1,95 @@
+authors:
+  - jheyy
+hero: "image/Dyx9FwYgMyNqy1kMGx8Orz6q0qC3/kogMUEiTkiI5gKqaxrsJ.jpg"
+alt: "Layered mountains."
+tags:
+  - css
+  - html
+date: 2022-08-18
+---
+
+The top layer sits above its related `document` in the browser viewport, and each `document` has one associated top layer. This means that elements promoted to the top layer needn't worry about `z-index` or DOM hierarchy. They also get a neat `::backdrop` pseudo-element to play with. The [Fullscreen API](https://fullscreen.spec.whatwg.org/#new-stacking-layer) spec goes into more details as [Fullscreen](https://developer.mozilla.org/docs/Web/API/Element/requestFullScreen) was a great example of the top layer in use before [`dialog` support](https://caniuse.com/dialog) came along.
+
+The top layer helps solve the problem of rendering content above the rest of the `document`.
+
+The important things to remember:
+- Top layer is outside of the `document` flow.
+- `z-index` has no effect in the top layer.
+- Each element in the top layer has a styleable `::backdrop` pseudo-element.
+- Each element and `::backdrop` generates a new stacking context.
+- Elements in the top layer are stacked in the order they appear in the set. The last one in, appears on top. If you want to promote an element, remove it, and add it again.
+
+
+How have we mimicked the top layer until now? Well, it's not uncommon to see developers dropping an empty container element at the end of the `body`. And then this will get used as a "faux" top layer. The idea is that this container gets positioned above everything else in the stack. When you want to promote something above everything else, you append it to that container. We can see this in popular packages like [SweetAlert](https://github.com/t4t5/sweetalert), [reactjs-popup](https://github.com/yjose/reactjs-popup), [Magnific Popup](https://github.com/dimsemenov/Magnific-Popup), and others. 
+
+{% Codepen {
+    user: 'web-dot-dev',
+    id: 'xxWYaor',
+    height: 450,
+    tab: 'result'
+  }
+%}
+
+
+With new built-in components and APIs like `dialog` and `pop-up`, you won't need to resort to these workarounds. You can promote content to the top layer. 
+
+UI frameworks allow us to co-locate promoted elements with their component counterparts. But, they often get separated in the DOM when it comes to rendering.
+
+By using the top layer, promoted elements are where you put them in your source code (for example, a `dialog`). It doesn't matter how many layers down in the DOM the element is. It will get promoted to the top layer and you can inspect it where you expect it to be, co-located with your component HTML. This makes it easier to inspect both the trigger element and the promoted element in the DOM at the same time. Particularly useful if your trigger element is making attribute updates, for example. This also has an added benefit for accessibility now that the elements are co-located.
```

Comment:
```suggestion
By using the top layer, promoted elements are where you put them in your source code (for example, a `<dialog>`). It doesn't matter how many layers down in the DOM the element is. It will get promoted to the top layer and you can inspect it where you expect it to be, co-located with your component HTML. This makes it easier to inspect both the trigger element and the promoted element in the DOM at the same time. Particularly useful if your trigger element is making attribute updates, for example. This also has an added benefit for accessibility now that the elements are co-located.
```